### PR TITLE
Added cosmic rate calculation tool

### PR DESCRIPTION
**Pull request by skeshri**
Added cosmic rate calculation tool for tracker alignment.

I have tested it locally before creating this pull request. Experts (@ghellwig and @mschrode) were monitoring this update.

### CODE REVIEW

**Comment by mmusich** [comment]
please test

File: Alignment/TrackerAlignment/plugins/CosmicRateAnalyzer.cc [hunk 1]
```diff
@@ -0,0 +1,529 @@
...
+  clustercollectionToken_(consumes<edmNew::DetSetVector<SiStripCluster> >(iConfig.getParameter<edm::InputTag>("tracks"))),
+  muonTags_(consumes<reco::MuonCollection>(iConfig.getParameter<edm::InputTag>("muons")))
+{
+   //now do what ever initialization is needed
+   //
+    treeEvent = fs->make<TTree>("Event","");
+    treeRun = fs->make<TTree>("Run","");
+    treeCluster = fs->make<TTree>("Cluster","");
+
+}
```
Comment (ghellwig) [inline Alignment/TrackerAlignment/plugins/CosmicRateAnalyzer.cc]: @skeshri: you are using the TFileService, i.e. you have to add `usesResource(TFileService::kSharedResource);` in the constructor
in addition you need to add the template parameter `edm::one::SharedResources` to the `edm::one::EDAnalyzer` from which you inherit

File: Alignment/TrackerAlignment/plugins/BuildFile.xml [hunk 2]
```diff
@@ -35,3 +35,23 @@
   <use   name="root"/>
   <flags EDM_PLUGIN="1"/>
 </library>
+
+<library file="CosmicRateAnalyzer.cc" name="AlignmentCosmicRateAnalyzerPlugins">
+  <use   name="Alignment/TrackerAlignment"/>
```
Comment (ghellwig) [inline Alignment/TrackerAlignment/plugins/BuildFile.xml]: @skeshri this is not needed, I think. From the cc file I cannot see that you use something from this package.

File: Alignment/TrackerAlignment/plugins/CosmicRateAnalyzer.cc [hunk 3]
```diff
@@ -0,0 +1,529 @@
...
+/*
+void 
+CosmicRateAnalyzer::endLuminosityBlock(edm::LuminosityBlock const&, edm::EventSetup const&)
+{
+}
+*/
+
+// ------------ method fills 'descriptions' with the allowed parameters for the module  ------------
+void
+CosmicRateAnalyzer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
```
Comment (ghellwig) [inline Alignment/TrackerAlignment/plugins/CosmicRateAnalyzer.cc]: @skeshri please add a suitable `fillDescriptions` method

File: Alignment/TrackerAlignment/test/cosmicRateAnalyzer_cfg.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+readFiles = cms.untracked.vstring()
+
+
+readFiles = cms.untracked.vstring( FileUtils.loadListFromFile (os.environ['CMSSW_BASE']+'/src/Alignment/TrackerAlignment/test/'+'fileList.txt') )
+process.source = cms.Source("PoolSource",
+			   fileNames = readFiles,
+			   )
+
+process.TFileService = cms.Service("TFileService", fileName = cms.string("Cosmic_rate_tuple.root") )
+process.demo = cms.EDAnalyzer("CosmicRateAnalyzer",
```
Comment (ghellwig) [inline Alignment/TrackerAlignment/test/cosmicRateAnalyzer_cfg.py]: @skeshri when you have added the filldescriptions method you can load the auto-generated file here and modify only the parameters you want

File: Alignment/TrackerAlignment/test/cosmicRateAnalyzer_cfg.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+#				tracks = cms.InputTag("ctfWithMaterialTracksP5"),                # Track collection for prompt RECO Dataset
+				tracks = cms.InputTag("ALCARECOTkAlCosmicsCTF0T"),		 # Track collection for stream and prompt ALCARECO Dataset
+				muons = cms.InputTag("muons1Leg"),		 # for muon Trigger timing information  
+#				fileName = cms.untracked.string("Cosmic_rate_tuple.root")
```
Comment (ghellwig) [inline Alignment/TrackerAlignment/test/cosmicRateAnalyzer_cfg.py]: @skeshri remove this comment because you use the file service

**Comment by ghellwig** [comment]
@skeshri please also link a presentation of your tool in the description of this pull request

**Comment by mmusich** [comment]
please test

**Comment by skeshri** [comment]
Hi I couldn't find any log at the link https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-15150/13959/git-merge-result

**Comment by mmusich** [comment]
@skeshri gitHub complains with: `This branch has conflicts that must be resolved`.
A rebase is likely needed

**Comment by mmusich** [comment]
tip: try to merge your PR with `git cms-merge-topic 15150` in a fresh IB. That should show what went wrong.

**Comment by skeshri** [comment]
I tried  git cms-merge-topic 15150. I got the following : 
From .
 * [new ref]                    -> merge-attempt
From https://github.com/cms-sw/cmssw
 * [new ref]         refs/pull/15150/head -> cms-sw/refs/pull/15150/head
Switched to branch 'merge-attempt'
Already up-to-date.
Switched to branch 'cosmic-rate-tool'
>> Checking Alignment/TrackerAlignment CMSSW_8_0_7_patch1
   x Alignment/TrackerAlignment/plugins/BuildFile.xml
Checking out these packages: 0

After that when I did git branch I got following : 
  CMSSW_8_0_X
  cms-sw/refs/pull/15150/head
* cosmic-rate-tool
  from-CMSSW_8_0_7_patch1

But I don't understand what exactly the problem is?

**Comment by skeshri** [comment]
there is some problem in format.

these are four branches :
CMSSW_8_0_X
cms-sw/refs/pull/15150/head
cosmic-rate-tool 
from-CMSSW_8_0_7_patch1

**Comment by mmusich** [comment]
@skeshri
```
cmsrel CMSSW_8_1_X_2016-07-10-2300
cd CMSSW_8_1_X_2016-07-10-2300/src/
cmsenv
git cms-merge-topic 15150
```

results in:
```
Auto-merging Alignment/TrackerAlignment/plugins/BuildFile.xml
CONFLICT (content): Merge conflict in Alignment/TrackerAlignment/plugins/BuildFile.xml
Automatic merge failed; fix conflicts and then commit the result.
Unable to merge branch refs/pull/15150/head from repository cms-sw.
```

you have an issue in: 

`Alignment/TrackerAlignment/plugins/BuildFile.xml`

**Comment by skeshri** [comment]
@mmusich 
I have removed only one line from BuildFile.xml as it was not needed. You can see at  https://github.com/cms-sw/cmssw/pull/15150/commits/538cea4a3956530bfe1020672cfe74da38f7c072
But I couldn't trace the exact problem.

**Comment by mmusich** [comment]
@skeshri try to rebase on top of the HEAD of CMSSW_8_1_X. It will fail the merge. You'll need to manually edit `Alignment/TrackerAlignment/plugins/BuildFile.xml` to remove the conflict symbols.
Once done commit and push again to the same branch.
Otherwise close this one and start from a fresh installation.

**Comment by skeshri** [comment]
@mmusich 
Thanks Marco. I guess fresh installation is easier for me. I'm closing this pull request.

**Pull request closed without merging**